### ISSUE TITLE
EntityKnockbackByEntityEvent called with wrong entities

### ISSUE THREAD

**Issue by vytskalt** [PaperSpigot/Paper#7168]
### Expected behavior

The EntityKnockbackByEntityEvent#hitBy() entity should be called with the entity that did the knockback

### Observed/Actual behavior

Instead, if the entity that did the knockback is a projectile or a TNT, the #hitBy entity is a player (shooter of the projectile or the one that lit the TNT) instead of the entity that actually did the knockback

### Steps/models to reproduce

Example 1:
1. Throw a snowball
2. EntityKnockbackByEntityEvent will be called with #hitBy set to the shooter of the snowball instead of the actual snowball entity

Example 2:
1. Get a TNT and flint and steel
2. Prime the TNT
3. EntityKnockbackByEntityEvent will be called with #hitBy set to the player that primed the TNT instead of the TNT entity

Example 3:
1. /summon tnt
2. EntityKnockbackByEntityEvent will not be called because there was no entity that primed the TNT

### Plugin and Datapack List

Only a plugin that was printing the #hitBy entity into the console. No datapacks.

### Paper version

This server is running Paper version git-Paper-85 (MC: 1.18.1) (Implementing API version 1.18.1-R0.1-SNAPSHOT) (Git: df8d28a)
You are running the latest version

**Comment by Machine-Maker** [comment]
I'm pretty sure this is intentional. The game tracks the source entity for the damage sources you named, tnt being ignited by a player or a projectile fired from a player/entity. 

That being said, I'm open to a nullable method in that event to get the actual entity. It should've been the actual entity before cause there are methods to get the "shooter" or a projectile if you actually wanted a reference to that entity, but I don't think it can be changed now.

**Comment by vytskalt** [comment]
> I'm pretty sure this is intentional. The game tracks the source entity for the damage sources you named, tnt being ignited by a player or a projectile fired from a player/entity.
> 
> That being said, I'm open to a nullable method in that event to get the actual entity. It should've been the actual entity before cause there are methods to get the "shooter" or a projectile if you actually wanted a reference to that entity, but I don't think it can be changed now.

That nullable method would still not fix the issue where the event isn't called if there's no source entity.
I don't see any way of fixing this issue without breaking plugins or creating a new event.

**Comment by Machine-Maker** [comment]
Yeah, so that's really a separate issue. You've got two issues here, one where you can't get the projectile if it has a shooter, and one where its tnt that didnt come from a player in some way. (I *think* that's the only was an entity can create knockback without having a source mob)

**Comment by vytskalt** [comment]
> Yeah, so that's really a separate issue. You've got two issues here, one where you can't get the projectile if it has a shooter, and one where its tnt that didnt come from a player in some way. (I _think_ that's the only was an entity can create knockback without having a source mob)

Well, they're both related so I made an issue for both of them. I can create a separate issue if you want though.

**Comment by Joo200** [comment]
Paper has a changed behavior from Spigot regarding the EntityKnockbackByEntityEvent from Bukkit. This event is deprecated for Paper but this event shouldn't behave different than on Spigot.

The method `org.bukkit.event.entity.EntityKnockbackByEntityEvent#getSourceEntity()` returns different entities on Spigot than on Paper:

Used code:
```java
    @EventHandler(ignoreCancelled = true)
    public void onEntityKnockbackByEntity(EntityKnockbackByEntityEvent event) {
        Entity damager = event.getSourceEntity();

        WorldGuardPlugin.inst().getLogger().info("Knockback from damager: " + damager.getClass().getCanonicalName());
    }
```

Spigot will call two events. I'm not sure why there are two events called.
```
[16:42:08] [Server thread/INFO]: [WorldGuard] Knockback from damager: org.bukkit.craftbukkit.v1_21_R1.entity.CraftPlayer
[16:42:08] [Server thread/INFO]: [WorldGuard] Knockback from damager: org.bukkit.craftbukkit.v1_21_R1.entity.CraftWindCharge
```

Paper calls only one event:
```
[16:46:42 INFO]: [WorldGuard] Knockback from damager: org.bukkit.craftbukkit.entity.CraftPlayer
```

If I want to cancel only knockbacks created by exploded windcharges I can't get the source entity correctly. Using Spigot and cancel for entities instance of AbstractWindCharge it works fine. On paper this doesn't work.

**Comment by lynxplay** [comment]
Well we cannot really change the returned value of our own events, eventho I would *somewhat* agree that `getPushedBy` is somewhat missleading to return the causing entity and not the direct one.

Fixing this for legacy events however is certainly something we should be doing.

**Comment by electronicboy** [comment]
The problem here is the disambiguation between what vanilla considers the causing entity and what bukkit considers the causing entity, and the widening gap between the two. part of the reality is that we just need to expose the real full context as much as possible and let people figure what they're after there

**Comment by Joo200** [comment]
Opened PR #11176 to fix the Bukkit behavior. More API in the paper events would be great but that would be a new feature. With more to do (and a duplicate of #7170 .

**Comment by paaulhier** [comment]
Will this be released in the next version?

**Comment by joelra** [comment]
I develop and maintain a land-claim plugin. This issue blocks my migration away from a deprecated event. I want to give a clear example of why the fix in #13472 is important and should be considered for addressing this reported issue.

**My use case**

My plugin controls knockback on players inside a claim. The plugin must stop knockback from TNT and creepers. The plugin must permit knockback from wind charges. Wind charge movement is a core game mechanic since 1.21. If the plugin stops it, players cannot use wind charges for movement.

**The problem**

To make the correct decision, the plugin must know the entity that caused the push. The modern event does not give this entity. `getHitBy()` returns the player who threw the projectile. It does not return the projectile.

Example 1: a player throws a wind charge. The event gives the player.
Example 2: a player ignites TNT. The event also gives the player.

The two events look the same. The plugin cannot see a difference.

**The current workaround**

The old `org.bukkit.event.entity.EntityKnockbackByEntityEvent` gives the correct entity. My plugin must use this old event. But this event is deprecated with `forRemoval = true`. The server shows a warning for my plugin at each start. When Paper removes the old event, plugins with this use case will have no correct solution.

**The server already has the data**

`CraftEventFactory.callEntityKnockbackEvent` receives both entities today. It puts the direct entity into the old event. It drops the direct entity from the modern event. The call sites need no changes. The missing piece is only a slot on the modern event.

**The history**

PR #7170 proposed this fix in December 2021. It didn't seem that fix was opposed. Instead feedback was about API design: the getter must not be nullable ([comment](https://github.com/PaperMC/Paper/pull/7170#issuecomment-1029312225)), and the names must be more descriptive, with `getHitBy()` deprecated ([comment](https://github.com/PaperMC/Paper/pull/7170#issuecomment-2254180753)). The PR became inactive and closed after three years. #13472 is the new attempt. This issue is now four years old.

**The fix**

PR #13472 adds `getDirectHitter()` to the modern event. This method returns the projectile. This is the data my use case needs. I support the naming feedback from #7170 which was: deprecate `getHitBy()` and choose clear names. With any clear name, I can move to the modern event and remove the deprecated listener.

I'd like to ask for a review from the main team on #13472, or update it with the desired names, and merge it. I'm happy to open a new PR if desired. Also I am available to test the change.
